astropy/cosmology/flrw.py

```diff
@@ -178,37 +184,26 @@ def __init__(self, H0, Om0, Ode0, Tcmb0=0.0*u.K, Neff=3.04, m_nu=0.0*u.eV,
             # It is worth the effort to keep track of massless ones separately
             # (since they are quite easy to deal with, and a common use case
             # is to set only one neutrino to have mass)
-            m_nu = self._m_nu
-            if m_nu.isscalar:
-                # Assume all neutrinos have the same mass
-                if m_nu.value == 0:
-                    self._nmasslessnu = self._nneutrinos
-                    self._nmassivenu = 0
-                else:
+            m_nu = self._m_nu  # has units eV
+            if np.min(m_nu.value) < 0:
+                raise ValueError("invalid (negative) neutrino mass encountered.")
+            elif m_nu.isscalar:  # Assume all neutrinos have the same mass.
+                if m_nu.value != 0.0:  # already dealt with m <= 0
                     self._massivenu = True
-                    self._nmasslessnu = 0
+                    self._massivenu_mass = np.full(self._nneutrinos, m_nu.value)
                     self._nmassivenu = self._nneutrinos
-                    self._massivenu_mass = (m_nu.value *
-                                            np.ones(self._nneutrinos))
-            else:
-                # Make sure we have the right number of masses
-                # -unless- they are massless, in which case we cheat a little
-                if m_nu.value.min() < 0:
-                    raise ValueError("Invalid (negative) neutrino mass"
-                                     " encountered")
-                if m_nu.value.max() == 0:
-                    self._nmasslessnu = self._nneutrinos
-                    self._nmassivenu = 0
-                else:
-                    self._massivenu = True
-                    if len(m_nu) != self._nneutrinos:
-                        errstr = "Unexpected number of neutrino masses"
-                        raise ValueError(errstr)
-                    # Segregate out the massless ones
-                    self._nmasslessnu = len(np.nonzero(m_nu.value == 0)[0])
-                    self._nmassivenu = self._nneutrinos - self._nmasslessnu
-                    w = np.nonzero(m_nu.value > 0)[0]
-                    self._massivenu_mass = m_nu[w]
+                    self._nmasslessnu = 0
+            elif len(m_nu) != self._nneutrinos:  # not scalar, check number of masses
+                raise ValueError("unexpected number of neutrino masses — "
+                                 f"expected {self._nneutrinos}, got {len(m_nu)}.")
+            # elif m_nu.value.max() == 0:
+            #     pass  # if this is true, min = max = 0 = default case (above)
+            elif m_nu.value.max() != 0:  # Different masses.
```

Comment:
This could be `any(m_nu.value > 0)` for clarity.

I think it might be clearer to put that further up top, and then do the check on scalar inside.  I.e., something like
```
if m_nu.shape not in ((), (self._nneutrinos,)):
    raise ValueError(...)
if any(m_nu.value < 0:
    raise ...
if any(m_nu.value > 0):
    if m_nu.isscalar:
        ...
    else:
        ...
```
But not a big deal if not.

Reply:
> This could be ``any(m_nu.value > 0)`` for clarity.

Much better.

> I think it might be clearer to put that further up top, and then do the check on scalar inside

I agree. I actually have a followup PR, dependent on this one, that cleans up more stuff. I'd prefer to implement this improvement in that PR. Thanks!